### PR TITLE
[RFC][WIP] stm32(L1x) RTC timer driver for power management

### PR DESCRIPTION
**Pull request by noelleclement**
WIP for STM32 RTC timer driver, **comments highly requested**!

Eventually should be split up into different PRs, but for feedback and testing I included everything in 1 PR for now.

### Included in this PR are:
- New STM32 RTC timer driver, which can be used as the kernel system clock, and sets wake-up alarms (Alarm A, not Periodic Wake Up) for exiting low-power modes. Subsecond alarm support is included to increase granularity to <1 sec.
- A STM32L152XC SoC port, since the current chosen/configured SoC for the stm32l1_disco board is CONFIG_SOC_STM32L151XB=y, which doesn't reflect reality and causes problems. The STM32L151XB SoCs are Cat. 1 SoCs, which don't support subsecond alarms, which trickles down to the STM32 HAL where it causes the SubSecond functions to not be included during compilation. 
- STM32L1x power management (power.c file) - currently only for stop mode (CPU shut off, context kept, some peripherals still available such as RTC)
- Sample application in which a thread (in loop) is put to sleep for 15500 msec, and 156 msec. Created to trigger the power management subsystem when kernel is idle, so this functionality specifically can be tested. 
- Clock reconfiguration (temporary) patch to allow reconfiguration after stop mode exit. See #28572.

### Background info:
With this driver STM32 SoCs that don't have LPTIM should still be able to use the power management subsystem, with low-power modes where context is kept (and kernel needs to resume post wake-up). Currently no solution is available to allow this.

For STM32Lx SoCs the SoC system clocks during run mode (HSI/HSE/MSI) aren't available during low-power modes (such as stop mode, where context is kept). If one would want to be able to wake-up from such a power mode, use of the RTC (LSI/LSE) is a common solution to generate the wake-up interrupt. However, if after wake-up the kernel needs to resume from the point before sleep, the elapsed time needs to be announced to the kernel system clock. 
A solution has been tested before where the kernel system clock was the same as the SoC system clock (the default cortex_m_systick, configured with HSI as clock source, which shuts off during stop mode). This was combined with the use of the STM32 RTC counter driver (to set the wake-up interrupt and provide a time reference post wake-up), but this caused some issues when the kernel was supposed to resume work post wake-up. This was most probably due to the kernel system clock having a different frequency from the RTC and the post wake-up announcement not being earlier than the 'regular' timer's time announcement. 
By creating a separate RTC timer driver only one frequency is used for the kernel timing source, which is in line with the proposal [[here](https://github.com/zephyrproject-rtos/zephyr/issues/30590#issuecomment-742606280)]. It also created the opportunity to tweak the driver's functionality to fit the requirements better, which would have been difficult if the counter driver would be adapted (due to the API requirements etc). This made it possible to increase granularity as well (by adding subsecond support).

### Additional info:
- Atm tested with stm32l1_disco board (& will start testing on our custom board soon too, which has the STM32L151CC SoC). 
- Haven't been able to prove that the stop mode is truly reached (power consumption is low enough) during testing with the stm32l1_disco board, hopefully will be able to do that with our custom board. 
- No additional support currently added for non-L1x SoCs, and only support for non-Cat. 1 L1x SoCs (which support subseconds)
- Support for both LSI and LSE RTC clock source (configuration dependent on choice, automatic config selections included)
- Atm CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC not set/used in the driver itself, since it might interfere with other drivers that count on that value reflecting the SoC system clock frequency, not the kernel system clock source frequency.

\\\

This is still very much WIP (see all the TODOs and commented code :) ), and not ready to merge, but I would highly appreciate comments on the following:
- Are timing edge-cases handled properly / how can I test this? (!!! most important atm !!!)
- Can this driver cause problems with the STM32 RTC **counter** driver? How should this be handled?
- Additional comments on overall functionality / configuration options / style

### CODE REVIEW

**Comment by noelleclement** [comment]
@pabigot If you have time, could you take a look a this?

**Comment by erwango** [comment]
@noelleclement Thanks for pushing this work! In order to ease the review, could you refactor your commits per impacted subsets (dts, clock control drivers, timer driver, samples) ?
So basically, you'd need to squash all your commits (you'll lose the steps of implementation, but from review point of view, what matters is the last state of things) and split into commits based on impacted subsystems and logical blocks.
This might help identifying some parts that could actually be merged sooner (as less controversial like dts refactoring parts for instance) and focus on the parts that are more subject to discussion.

**Comment by noelleclement** [comment]
@erwango Will do!

File: samples/boards/stm32/stm32l1_disco/rtc_systemclock/README.rst [hunk 1]
```diff
@@ -0,0 +1,88 @@
+TODOOOOOOOO (if you read this I definitely forgot it)
+
+.. _stm32-stm32l1_disco-rtc_systemclock-sample:
+
+stm32l1_disco RTC system clock demo
```
Comment (pabigot) [inline samples/boards/stm32/stm32l1_disco/rtc_systemclock/README.rst:5]: Is this specific to the STM32L1 or will it work with multiple STM32 RTC peripherals?

**Review by pabigot — COMMENTED** [review]
I don't have a lot of input on this because I'm not familiar with the RTC. What's the maximum TICKS_PER_SEC that can be supported?  If it isn't at least 100 Hz some drivers might have problems, because that was the old standard.  Currently most boards will use 10 kHz.

**Comment by noelleclement** [comment]
Had to create a new PR to implement te suggested commit-change: #30790

**Pull request closed without merging**